### PR TITLE
Removing Http2StreamRemovalPolicy

### LINKED ISSUE #3448: Get rid of Http2StreamRemovalPolicy
The `DefaultHttp2StreamRemovalPolicy` by default waits 5 seconds after a stream has been closed before removing it from the connection.  The reason for this is to support the following case from the HTTP2 spec (From [Section 5.1](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-5.1)):

    "WINDOW_UPDATE or RST_STREAM frames can be received in this state
     for a short period after a DATA or HEADERS frame containing an
     END_STREAM flag is sent.  Until the remote peer receives and
     processes RST_STREAM or the frame bearing the END_STREAM flag, it
     might send frames of these types.  Endpoints MUST ignore
     WINDOW_UPDATE or RST_STREAM frames received in this state, though
     endpoints MAY choose to treat frames that arrive a significant
     time after sending END_STREAM as a connection error
     of type PROTOCOL_ERROR."

These cases don't really seem to warrant the extra complexity that DefaultHttp2StreamRemovalPolicy introduces.  It would be better to just always ignore WINDOW_UPDATE and RST_STREAM for non-existent streams.  This way we can always remove the streams as soon as they're closed to avoid potential memory leaks.

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

Due to a recent flurry of cleanup and fixes, we no longer need the stream removal policy to protect against recently removed streams. We should get rid of it.

Modifications:

Removed Http2StreamRemovalPolicy and everywhere it's used.

Result:

Fixes #3448

### CODE REVIEW

**Comment by nmittler** [comment]
@Scottmitch @buchgr I think this is straight-forward.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 1]
```diff
@@ -962,28 +940,28 @@ private boolean isLocal() {
...
-            /**
-             * Trigger the original intention of this event. Expect to modify {@link #streams}.
-             */
-            void process();
-        }
+        void process();
+    }
+
+    /**
+     * Default implementation of the {@link ActiveStreams} class.
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Kind of strange to say "Default implementation" as this is not an implementation of an interface.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 2]
```diff
@@ -1060,8 +1038,8 @@ void removeFromActiveStreams(DefaultStream stream) {
                         listeners.get(i).onStreamClosed(stream);
                     }
                 } finally {
-                    // Mark this stream for removal.
-                    removalPolicy.markForRemoval(stream);
+                    // Remove the stream entirely if it has no children in the priority tree.
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @nmittler I think this comment is not right. Streams should be removed even if they have children, but the children should be reprioritised and get a new parent?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @buchgr I believe the comment accurately reflects the current behavior

@Scottmitch any thoughts?

Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @nmittler hmm okay, would this then not allow for leaks i.e. "parent streams" sticking around after all their children have been removed?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: +1. I removed the interface and just kept the class as part of review.  A comment similar to what you added for `interface Event` may make more sense?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: > @buchgr I believe the comment accurately reflects the current behavior
+1.

The current behavior is such that streams remain in the priority tree until `prioritizableForTree() == 0`.  Events which can change this count (parent change, stream removal from tree, state change, etc...) re-evaluate whether the stream still belongs in the tree, and calls `removeStream` if not. I'm not sure we need this comment here though...`removeStream` has a pretty detailed comment on it...maybe just clean that comment up if it is not adequite?

**Comment by Scottmitch** [comment]
@nmittler - Good stuff. Just one follow-up point on @buchgr's comment. BTW I'm a big fan of PRs which have a add/removal ratio meter that is all red :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: > The current behavior is such that streams remain in the priority tree until prioritizableForTree() == 0. Events which can change this count (parent change, stream removal from tree, state change, etc...) re-evaluate whether the stream still belongs in the tree, and calls removeStream if not.

nice. thanks for clarifying!

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler remove code for the win! :beers:

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Comment removed.

**Comment by nmittler** [comment]
@Scottmitch @buchgr any other comments before I cherry-pick?

**Comment by buchgr** [comment]
@nmittler nope. LGTM!

**Comment by Scottmitch** [comment]
@nmittler - LGTM!  Thanks!

**Comment by nmittler** [comment]
cherry-picked into 4.1 and master.

**Pull request closed without merging**